File: interview questions/breadthFirstSearch.py
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))
        return self

    def breadthFirstSearch(self, array):
        queue = [self]

        while len(queue) > 0:
            node = queue.pop(0)
            array.append(node.name)
            queue.extend(node.children)
        return array

File: interview questions/breadthFirstSearch.js
class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

  breadthFirstSearch(array) {
    let queue = [this];

    while (queue.length > 0) {
      let node = queue.shift();
      array.push(node.name);

      for (let i = 0; i < node.children.length; i++) {
        let childNode = node.children[i];
        queue.push(childNode);
      }
    }

    return array;
  }
}
